polkadot/xcm/pallet-xcm/src/precompiles.rs

```diff
@@ -0,0 +1,160 @@
+use tracing::error;
+use xcm::MAX_XCM_DECODE_DEPTH;
+use xcm_executor::traits::WeightBounds;
+
+alloy::sol!("src/precompiles/IXcm.sol");
+use IXcm::*;
+
+pub struct XcmPrecompile<T>(PhantomData<T>);
+
+impl<Runtime> Precompile for XcmPrecompile<Runtime>
+where
+	Runtime: crate::Config + pallet_revive::Config,
+{
+	type T = Runtime;
+	const MATCHER: AddressMatcher = AddressMatcher::Fixed(NonZero::new(10).unwrap());
+	const HAS_CONTRACT_INFO: bool = false;
+	type Interface = IXcm::IXcmCalls;
+
+	fn call(
+		_address: &[u8; 20],
+		input: &Self::Interface,
+		env: &mut impl Ext<T = Self::T>,
+	) -> Result<Vec<u8>, Error> {
+		let origin = env.caller();
+		let frame_origin = match origin {
+			Origin::Root => frame_system::RawOrigin::Root.into(),
+			Origin::Signed(account_id) =>
+				frame_system::RawOrigin::Signed(account_id.clone()).into(),
+		};
+
+		match input {
+			IXcmCalls::xcmSend(IXcm::xcmSendCall { destination, message }) => {
+				let _ = env.charge(<Runtime as Config>::WeightInfo::send())?;
+
+				let final_destination = VersionedLocation::decode_all(&mut &destination[..])
+					.map_err(|error| {
+						error!(target: "xcm::precompiles", ?error, "XCM send failed: Invalid destination format");
+						Error::Revert("XCM send failed: Invalid destination format".into())
+					})?;
+
+				let final_message = VersionedXcm::<()>::decode_all_with_depth_limit(
+					MAX_XCM_DECODE_DEPTH,
+					&mut &message[..],
+				)
+				.map_err(|error| {
+					error!(target: "xcm::precompiles", ?error, "XCM send failed: Invalid message format");
+					Error::Revert("XCM send failed: Invalid message format".into())
+				})?;
+
+				crate::Pallet::<Runtime>::send(
+					frame_origin,
+					final_destination.into(),
+					final_message.into(),
+				)
+				.map(|message_id| message_id.encode())
+				.map_err(|error| {
+					error!(
+						target: "xcm::precompiles",
+						?error,
+						"XCM send failed: destination or message format may be incompatible"
+
+					);
+					Error::Revert(
+						"XCM send failed: destination or message format may be incompatible".into(),
+					)
+				})
+			},
+			IXcmCalls::xcmExecute(IXcm::xcmExecuteCall { message, weight }) => {
+				let weight = Weight::from_parts(weight.refTime, weight.proofSize);
+				let charged_amount = env.charge(weight)?;
+
+				let final_message = VersionedXcm::decode_all_with_depth_limit(
+					MAX_XCM_DECODE_DEPTH,
+					&mut &message[..],
+				)
+				.map_err(|error| {
+					error!(target: "xcm::precompiles", ?error, "XCM execute failed: Invalid message format");
+					Error::Revert("Invalid message format".into())
+				})?;
+
+				crate::Pallet::<Runtime>::execute(frame_origin, final_message.into(), weight)
+					.map(|post_dispatch_info| {
+						// Adjust charged amount to account for the actual weight consumed.
+						if let Some(actual_weight) = post_dispatch_info.actual_weight {
+							env.gas_meter_mut().adjust_gas(charged_amount.clone(), RuntimeCosts::Precompile(actual_weight));
+						}
```

Comment:
@athei Added the weight refund here and in the error case. Had to `.clone()` the charged weight. Is it reasonable?